### PR TITLE
Conversion of Parallel Stochastic Gradient Descent (SymSGD) to estimator

### PR DESCRIPTION
**Pull request by artidoro**
Ongoing work on converting the trainers to estimators. This PR converts the Parallel Stochastic Gradient Descent classification trainer (SymSGD).

### CODE REVIEW

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 1]
```diff
@@ -26,16 +25,14 @@
...
 namespace Microsoft.ML.Runtime.SymSgd
 {
-    using TPredictor = IPredictorWithFeatureWeights<Float>;
+    using TPredictor = IPredictorWithFeatureWeights<float>;
 
     /// <include file='doc.xml' path='doc/members/member[@name="SymSGD"]/*' />
-    public sealed class SymSgdClassificationTrainer :
-        TrainerBase<TPredictor>,
-        ITrainer<TPredictor>
+    public sealed class SymSgdClassificationTrainer : TrainerEstimatorBase<BinaryPredictionTransformer<TPredictor>, TPredictor>
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >TrainerEstimatorBase [](start = 54, length = 20)

SingleFeatureTrainerEstimatorBase

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 2]
```diff
@@ -26,16 +25,14 @@
...
     {
         public const string LoadNameValue = "SymbolicSGD";
         public const string UserNameValue = "Symbolic SGD (binary)";
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >public [](start = 8, length = 6)

internal, all 4 if you can.

**Comment by sfilipi** [comment]
>             public float? LearningRate;

might need to have params in the new constructor for this, and all args that don't have defaults. 

---
Refers to: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs:58 in 6697257. [](commit_id = 669725705de6ff7f3b3d932c47d86f7ef228575f, deletion_comment = False)

**Comment by sfilipi** [comment]
>             public float L2Regularization;

Might need to have a value for this on the new constructor; since no default. 

---
Refers to: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs:63 in 6697257. [](commit_id = 669725705de6ff7f3b3d932c47d86f7ef228575f, deletion_comment = False)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 3]
```diff
@@ -155,25 +152,69 @@ public override TPredictor Train(TrainContext context)
...
+        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
+        public SymSgdClassificationTrainer(IHostEnvironment env, string featureColumn, string labelColumn,
+            string weightColumn = null, Action<Arguments> advancedSettings = null)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), TrainerUtils.MakeR4VecFeature(featureColumn),
+                  TrainerUtils.MakeR4ScalarLabel(labelColumn), TrainerUtils.MakeR4ScalarWeightColumn(weightColumn))
+        {
+            var args = new Arguments();
+
+            //apply the advanced args, if the user supplied any
+            args.Check(Host);
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >args.Check(Host); [](start = 12, length = 17)

vice-versa

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 4]
```diff
@@ -155,25 +152,69 @@ public override TPredictor Train(TrainContext context)
...
+        /// <param name="labelColumn">The name of the label column.</param>
+        /// <param name="featureColumn">The name of the feature column.</param>
+        /// <param name="weightColumn">The name for the column containing the initial weight.</param>
+        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
+        public SymSgdClassificationTrainer(IHostEnvironment env, string featureColumn, string labelColumn,
+            string weightColumn = null, Action<Arguments> advancedSettings = null)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), TrainerUtils.MakeR4VecFeature(featureColumn),
+                  TrainerUtils.MakeR4ScalarLabel(labelColumn), TrainerUtils.MakeR4ScalarWeightColumn(weightColumn))
+        {
+            var args = new Arguments();
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >var args [](start = 12, length = 8)

you can use _args directly.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 5]
```diff
@@ -155,25 +152,69 @@ public override TPredictor Train(TrainContext context)
...
+            _args = args;
+            Info = new TrainerInfo();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="SymSgdClassificationTrainer"/>
+        /// </summary>
+        internal SymSgdClassificationTrainer(IHostEnvironment env, Arguments args)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), TrainerUtils.MakeR4VecFeature(args.FeatureColumn),
+                  TrainerUtils.MakeR4ScalarLabel(args.LabelColumn), TrainerUtils.MakeR4ScalarWeightColumn(null))
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >MakeR4ScalarLabel [](start = 31, length = 17)

bool

File: test/Microsoft.ML.Tests/TrainerEstimators/SymSgdClassificationTests.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            new TextLoader.Arguments()
+            {
+                HasHeader = true,
+                Column = new[]
+                {
+                    new TextLoader.Column("Label", DataKind.R4, 0),
+                    new TextLoader.Column("Features", DataKind.R4, new [] { new TextLoader.Range(1, 9) } )
+                }
+            }).Read(new MultiFileSource(GetDataPath(TestDatasets.breastCancer.trainFilename)));
+        }
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/TrainerEstimators/SymSgdClassificationTests.cs]: There is a GetBinaryPipeline() methof on TrainerEstimators. See if it helps reduce duplication.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 1 above)
Comment (artidoro) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Sorry I can't find that. what is the name of the base class I 
should derive from?

---
In reply to: [220056284](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220056284) [](ancestors = 220056284)

Comment (artidoro) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Why would this not work by the way?


---
In reply to: [220256710](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220256710) [](ancestors = 220256710,220056284)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 5 above)
Comment (sfilipi) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: You need a scalar bool label for binary classification. 

---
In reply to: [220057042](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220057042) [](ancestors = 220057042)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 1 above)
Comment (sfilipi) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: My bad, this is fine. got it confused with the TransformerPredictors 

---
In reply to: [220256891](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220256891) [](ancestors = 220256891,220256710,220056284)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 7]
```diff
@@ -155,25 +152,66 @@ public override TPredictor Train(TrainContext context)
...
+            string weightColumn = null, Action<Arguments> advancedSettings = null)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), TrainerUtils.MakeR4VecFeature(featureColumn),
+                  TrainerUtils.MakeBoolScalarLabel(labelColumn), TrainerUtils.MakeR4ScalarWeightColumn(weightColumn))
+        {
+            _args = new Arguments();
+
+            // Apply the advanced args, if the user supplied any.
+            _args.Check(Host);
+            advancedSettings?.Invoke(_args);
+            _args.FeatureColumn = featureColumn;
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >args.FeatureColumn [](start = 13, length = 18)

weights column

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 5 above)
Comment (artidoro) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Thanks for the catch, I had only modified it above and forgot this instance. 

---
In reply to: [220299186](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220299186) [](ancestors = 220299186,220057042)

File: test/Microsoft.ML.Tests/TrainerEstimators/SymSgdClassificationTests.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+using Xunit;
+
+namespace Microsoft.ML.Tests.TrainerEstimators
+{
+    public partial class TrainerEstimators
+    {
+        [Fact]
+        public void TestEstimatorSymSgdClassificationTrainer()
+        {
+            (IEstimator<ITransformer> pipe, IDataView dataView) = GetBinaryClassificationPipeline();
```
Comment (TomFinley) [inline test/Microsoft.ML.Tests/TrainerEstimators/SymSgdClassificationTests.cs]: >IEstimator<ITransformer> [](start = 13, length = 24)

For things like this don't be afraid of `var`.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 7 above)
Comment (TomFinley) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Still missing `weightColumn` a couple commits after @sfilipi's comment

---
In reply to: [220300485](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220300485) [](ancestors = 220300485)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 9]
```diff
@@ -155,25 +152,66 @@ public override TPredictor Train(TrainContext context)
...
 
-        public SymSgdClassificationTrainer(IHostEnvironment env, Arguments args)
-            : base(env, LoadNameValue)
+        /// <summary>
+        /// Initializes a new instance of <see cref="SymSgdClassificationTrainer"/>
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="labelColumn">The name of the label column.</param>
+        /// <param name="featureColumn">The name of the feature column.</param>
+        /// <param name="weightColumn">The name for the column containing the initial weight.</param>
```
Comment (TomFinley) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: >param name="weightColumn">The name for the column containing the initial weight.</param> [](start = 13, length = 88)

If we don't support weighted training this should be removed, at least in its description of a column.

I *think* what was actually intended, because it says "initial weight," was that when you train, an initial predictor from which you get initial weights should be supported. (e.g., as we see here:)

[https://github.com/dotnet/machinelearning/blob/893a3852bded9b8785b39e7973c02758069621b8/src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs#L145](https://github.com/dotnet/machinelearning/blob/893a3852bded9b8785b39e7973c02758069621b8/src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs#L145)

However this would be a parameter to the training/fitting method, not the constructor.

Comment (artidoro) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: That is actually already taken care of.
TrainerEstimatorBase has TrainTransformer method, which takes an initial predictor, and passes that to TrainModelCore through the trainContext

---
In reply to: [220342250](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220342250) [](ancestors = 220342250,220340515)

**Comment by artidoro** [comment]
>             public float L2Regularization;

After discussing with Pete, we should keep this as an advanced setting. The default value is 0 right now. 

---
In reply to: [424204987](https://github.com/dotnet/machinelearning/pull/1012#issuecomment-424204987) [](ancestors = 424204987)

---
Refers to: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs:63 in 6697257. [](commit_id = 669725705de6ff7f3b3d932c47d86f7ef228575f, deletion_comment = False)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 9 above)
Comment (TomFinley) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: You're talking about this I guess.

[https://github.com/dotnet/machinelearning/blob/d7b062de20c091cb4af5f58dda4bd0b98ad66c3e/src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs#L127-L128](https://github.com/dotnet/machinelearning/blob/d7b062de20c091cb4af5f58dda4bd0b98ad66c3e/src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs#L127-L128)

@artidoro, no, as you see this method is *protected*. It is not accessible to people that instantiate the class. So, no, not taken care of.

Your PR was already merged, so whoops, too late I guess. Maybe we can file an issue to fix this post 0.6.

---
In reply to: [220345387](https://github.com/dotnet/machinelearning/pull/1012#discussion_r220345387) [](ancestors = 220345387,220342250,220340515)
